File: enode/src/main/java/com/enode/infrastructure/impl/ManyType.java
package com.enode.infrastructure.impl;

import java.util.HashSet;
import java.util.List;
import java.util.stream.Collectors;

public class ManyType {

    private List<Class> types;

    public ManyType(List<Class> types) {
        if (new HashSet<>(types).size() != types.size()) {
            throw new IllegalArgumentException("Invalid ManyType:" + String.join("|", types.stream().map(Class::getName).collect(Collectors.toList())));
        }
        this.types = types;
    }

    public List<Class> getTypes() {
        return types;
    }

    @Override
    public int hashCode() {
        return types.stream().map(Object::hashCode).reduce((x, y) -> x ^ y).get();
    }

    @Override
    public boolean equals(Object obj) {

        if (obj == this) {
            return true;
        }

        if (obj == null) {
            return false;
        }

        if (!(obj instanceof ManyType)) {
            return false;
        }

        ManyType other = (ManyType) obj;

        if (this.types.size() != other.types.size()) {
            return false;
        }

        return types.stream().allMatch(type -> other.types.stream().anyMatch(x -> x == type))
                && other.types.stream().allMatch(type -> types.stream().anyMatch(x -> x == type));
    }
}

File: enode/src/main/java/org/enodeframework/infrastructure/impl/ManyType.kt
package org.enodeframework.infrastructure.impl

/**
 * @author anruence@gmail.com
 */
class ManyType(types: List<Class<*>>) {
    val types: List<Class<*>>

    override fun hashCode(): Int {
        return types.map { obj: Class<*> -> obj.hashCode() }.reduce { x: Int, y: Int -> x xor y }
    }

    override fun equals(other: Any?): Boolean {
        if (other === this) {
            return true
        }
        if (other == null) {
            return false
        }
        if (other !is ManyType) {
            return false
        }
        return if (types.size != other.types.size) {
            false
        } else types.any { type: Class<*> ->
            other.types.any { x: Class<*> -> x == type }
        } && other.types.all { type: Class<*> -> types.any { x: Class<*> -> x == type } }
    }

    init {
        require(HashSet(types).size == types.size) {
            String.format("Invalid ManyType: %s", types.joinToString("|") { obj: Class<*> -> obj.name })
        }
        this.types = types
    }
}